axlearn/common/trainer_config_modifier.py

```diff
@@ -277,3 +286,51 @@ def __call__(self, cfg: SpmdTrainer.Config) -> SpmdTrainer.Config:
         for config_modifier_fn in self._config_modifiers:
             cfg = config_modifier_fn(cfg)
         return cfg
+
+
+class FP8ConfigModifier(ConfigModifier):
+    """Update the trainer config to use FP8 training."""
+
+    @config_class
+    class Config(ConfigModifier.Config):
+        """Configure FP8ConfigModifier. See QuantizedDotGeneral.Config."""
+
+        fp8_amax_history_length: Required[int] = REQUIRED
+
+    def __call__(self, cfg: SpmdTrainer.Config) -> SpmdTrainer.Config:
+        """Override dense layer to use FP8 quantized dot and add gradient rules for FP8 stats."""
+        override_cfg: FP8ConfigModifier.Config = self.config
+        quantized_dot_general = QuantizedDotGeneral.default_config().set(
+            quantization_type=DotGeneralQuantizationType.FP_8,
+            fp8_amax_history_length=override_cfg.fp8_amax_history_length,
+        )
+
+        def visit_fn(_, value):
+            if isinstance(value, DenseGeneralBaseLayer.Config):
+                value.quantized_dot_general = quantized_dot_general
+
+        def enter_fn(_, value, default_kv):
+            return None if isinstance(value, DenseGeneralBaseLayer.Config) else default_kv
+
+        cfg.visit(visit_fn=visit_fn, enter_fn=enter_fn)
+
+        update_cfg: OverrideInplaceUpdateTransformation.Config = (
+            OverrideInplaceUpdateTransformation.default_config()
+        )
+        update_cfg.rules = [
+            f".*/{x}"
+            for x in [
+                "input_scale",
+                "kernel_scale",
+                "output_grad_scale",
+                "input_amax_history",
+                "kernel_amax_history",
+                "output_grad_amax_history",
+            ]
+        ]
+        transformation = maybe_instantiate(cfg.learner.optimizer)
```

Comment:
Usually we avoid instantiation when constructing configs to preserve the readability of configs. Can we defer the `maybe_instantiate` call to `OverrideInplaceUpdateTransformation.__init__`?

Reply:
Done.